DIPs/DIP1xxx-JS.md

```diff
@@ -0,0 +1,121 @@
+who's handling the pull requests. Standardizing the process makes sure that
+deprecations are done very publicly and carefully, so as to minimize breakage and
+to provide clear fixes for user code.
+
+## Description
+
+The following lays out the rules to follow to deprecate a part of D.
+
+A symbol or feature must not be marked for removal on a specific date, but rather on a
+specific release. This allows users to easily know if upgrading will break their
+code or not.
+
+Users must be given at least 10 non-patch releases before the deprecated features
+are removed. More releases can be given if the removed code is commonly used.
+There are two cases where the deprecation period is allowed to be shorter:
+
+1. The code or feature is notably dangerous or unsafe, and users need to remove
+it from their code as soon as possible.
+2. The existance of the current code precludes its own fix or the fix of an equally
+important issue.
+
+Shorting the deprecation period should be done with caution to avoid giving D
+an image of instability.
+
+At the time of the pull request for deprecation, all code in Phobos, Druntime,
+and DMD must be updated to remove use of the effected code. Any projects that
+are tested on the Project Tester that are broken should also have their
+maintainers notified.
+
+Both at the time of deprecation and removal, a changelog entry must be made. This
+changelog entry should have a short motivation for the deprecation (or removal)
+and should to describe what steps can be taken by the user to upgrade their codebase.
+
+In order to facilitate on schedule deprecations, a comment of the format
+`@@@DEPRECATED_[version]@@@` should be added to the top of the code to be removed/disabled.
+This comment allows code to be easily searched before every release to
+catch all planned deprecations.
+
+### Public Functions, Types, and Modules
+
+All removals (or changes that make the code `private`/`package`) of public functions,
+types, and modules must be accompanied with a deprecation period.
+
+The symbol(s) must be marked using the `deprecated` keyword with a message containing
+the planned removal period. A pointer to more information should also be added. E.g.
+"See the 2.080 changelog for more details" or "See the function documentation for more
+details". The documentation of the symbol(s) must be updated noting the
+deprecation and removal plan. The documentation should contain information to help
+the users using the symbol(s) transition their code away from the symbol(s).
+
+If the deprecation is occuring because the symbol(s) are being replaced by new
+symbols, both the old and the new symbol(s) should be availible un-deprecated
+in at least one release to allow users to build their code without issue on
+both the `stable` and `master` branches.
+
+On the fifth release in the deprecation period, the documentation for the symbol
+should be removed while keeping the code itself public until removal.
```

Comment:
Does it need to be that long?
Once a symbol is removed from the documentation, the biggest cost is paid as it only triggers the deprecation messages, but no longer affects the learning process of newcomers (or other people looking at the doc for help).
If the argument is for the old symbols being searchable via Google, then there are now the docarchives.dlang.io with an indexed snapshot of every release + the changelog entries on dlang.org

Many projects also simply undocument their deprecated symbols with their removal (and sometimes put them on a special deprecated list) which imho is a better way - especially given that we constantly seem to forget to undocument the symbols - but I guess that's too much to ask for :/

Reply:
Yes, and much as I hate to say it, I'm honestly sick of discussing this. This is the length that we came to after discussion on quite a few occasions and complaints both about it being too short and too long. We need to balance the ability to remove symbols that we don't want around anymore and not breaking existing code. Some folks in the past have pushed for _never_ removing deprecated symbols, and both Andrei and Walter have expressed opinions along those lines at one time or another. For instance, at one point, Walter went to update one of his programs, and it didn't compile anymore, because some symbols had been deprecated and removed since the last time he'd updated it, and he was very unhappy about it. Those symbols had gone through the full two year deprecation process. Imagine how much more complaining we'd get along those lines if the length of the deprecation process were short. Every time that the length of the deprecation process comes up for discussion, I'm worried that Walter or Andrei is going to put their foot down and insist that it's not a problem to keep symbols around forever, and I really don't want to see that happen.

Various devs (usually those who are constantly up-to-date with the compiler that they use and probably who use master most of the time) like the idea of quickly deprecating stuff and then removing it, whereas other folks aren't as on top of things, and for them, fast deprecations and removals cause them problems. And the fact that ldc or gdc are not always being caught up with where dmd is just make that worse. Those needs need to be balanced, and if anything, over time, as our user base has grown and the amount of D code that's been written has grown, the need to avoid breaking code has grown.

After trial and error, we came to the process of having symbols deprecated and documented for one year and then deprecated and undocumented for one year, and over all, that's worked fairly well. It balances the needs of being able to remove old symbols while not breaking code, and it balances the need for folks to look at the documentation for the symbols they're replacing and the need to discourage the use of such symbols by not having them documented.

Yes, if there's a symbol that we really want to get rid of, the length of the deprecation process can be annoying, but we shouldn't be deprecating symbols very frequently, because that's just going to make folks complain about how D isn't stable, and it's a sign that we keep screwing up what we put in Phobos. But at least we have a process for removing such symbols, even if it's not lightning fast. There are other languages, that tend to be far more rigid with their standard library, and some stuff sits there _forever_ even if it's deprecated.

I won't claim by any means that our deprecation process is perfect, but I think that how long it is and how we deal with the documentation balances things fairly well.

Reply:
FWIW I wasn't tackling the two-year deprecation period, but the period of how long the deprecated symbol still stays on dlang.org (I don't like the long deprecation period, but I wasn't proposing to change it).

> Some folks in the past have pushed for never removing deprecated symbols, and both Andrei and Walter have expressed opinions along those lines at one time or another. 

The State of D survey has different results: https://rawgit.com/wilzbach/state-of-d/master/report.html#180373330 (only 2% of the respondents said "never break code")

> After trial and error, we came to the process of having symbols deprecated and documented for one year and then deprecated and undocumented for one year, and over all, that's worked fairly well. 
> ...
> it balances the need for folks to look at the documentation for the symbols they're replacing and the need to discourage the use of such symbols by not having them documented.

Well, the only thing I was questioning whether we can remove the symbol earlier from the docs as two major things changed:
1) docarchives.dlang.io
2) We require visible changelog entries for every deprecation

> and it's a sign that we keep screwing up what we put in Phobos.

Well, two weeks ago I made a "small" list of things that I would change if we can do breaking changes in Phobos. Unfortunately it's a very long list and I stopped at point 135 and being not even through a quarter of Phobos's symbols. 

Anyhow I don't want to block this DIP with this, I just wanted to say that I personally feel that symbols should be undocumented sooner as they are still well visible for people searching for then and removing them helps to avoid encouragement of using deprecated symbols.

Reply:
Having access to older documentation helps, but if the symbols are undocumented too quickly, it becomes harder for folks updating their code to figure out what the old symbol did exactly so that they can properly update their code. Having them documented for a year and then undocumented for a year came out of trying to balance the desire to remove the documentation as quickly as possible so that folks definitely wouldn't use the symbol and the desire to keep the documentation around the entire time so that folks could easily find it to figure out how to update their code to not use it. There have been folks who have argued in favor of removing the documentation ASAP and others have argued for keeping it around longer. IIRC, at one point, Walter even wanted to have the documentation around _after_ the symbol was gone. The current approach was the compromise between the two extremes.

Now, we didn't use to have a way to access older documentation, so that does arguably change things, but I'm not sure that it's a good idea to speed up removing the documentation, especially since most folks are going to look at the main documentation, and I don't think that very many folks even know about docarchives.dlang.io. Having older versions of the documentation available has been discussed previously, but I think that this is the first that I've heard about there being anywhere that had it other than downloading the zip file for the specific compiler version. How would someone even know to look at docarchives.dlang.io to find older documentation?